migration/v0.43/apply/20-upgrade-opensearch.sh

```diff
@@ -0,0 +1,163 @@
+  log_info "- Getting name of .opensearch_dashboards index"
+  os_dashboards_index=$(curl -sS -kL -u "${user}:${password}" -X GET "${os_url}"/_alias/.opensearch_dashboards | jq -r 'to_entries | .[0].key')
+
+  if [[ "${os_dashboards_index}" != .opensearch_dashboards* ]]; then
+    log_fatal "Failed to get index name of the .opensearch_dashboards alias"
+  fi
+
+  log_info "- Marking index '${os_dashboards_index}' as read-only"
+  resp=$(curl -sS -kL -u "${user}:${password}" -X PUT "${os_url}"/"${os_dashboards_index}"/_settings -H 'Content-Type: application/json' -d'
+  {
+    "settings": {
+      "index.blocks.write": true
+    }
+  }
+  ')
+
+  acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+  if [ "${acknowledged}" = "true" ]; then
+    log_info "- Marked '${os_dashboards_index}' as read-only"
+  else
+    log_fatal "Failed to mark index '${os_dashboards_index}' as read-only" "${resp}"
+  fi
+
+  log_info "- Cloning index '${os_dashboards_index}' to index .kibana_1"
+  resp=$(curl -sS -kL -u "${user}:${password}" -X PUT "${os_url}"/"${os_dashboards_index}"/_clone/.kibana_1)
+  acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+  if [ "${acknowledged}" = "true" ]; then
+    log_info "- Successfully cloned '${os_dashboards_index}' to .kibana_1"
+  else
+    log_fatal "Failed to clone index '${os_dashboards_index}' to .kibana_1" "${resp}"
+  fi
+
+}
+
+disable_read_only_kibana_index() {
+
+  log_info "- Disabling read-only mode for .kibana_1"
+  resp=$(curl -sS -kL -u "${user}:${password}" -X PUT "${os_url}"/.kibana_1/_settings -H 'Content-Type: application/json' -d'
+  {
+    "settings": {
+      "index.blocks.write": false
+    }
+  }
+  ')
+  acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+  if [ "${acknowledged}" = "true" ]; then
+    log_info "- Successfully disabled read-only mode for .kibana_1"
+  else
+    log_fatal "Failed to disable read-only mode for .kibana_1" "${resp}"
+  fi
+
+}
+
+create_kibana_alias() {
+
+  resp=$(curl -sS -kL -u "${user}:${password}" -X GET "${os_url}"/_alias/.kibana)
+  if [[ $(echo "${resp}" | jq -r 'to_entries | .[0].key') == .kibana* ]]; then
+    log_info "- Alias .kibana already exists, skipping"
+  elif [[ $(echo "${resp}" | jq -r '.status') == "404" ]]; then
+    log_info "- Creating alias .kibana"
+    resp=$(curl -sS -kL -u "${user}:${password}" -X PUT "${os_url}"/.kibana_1/_aliases/.kibana)
+    acknowledged=$(echo "${resp}" | grep "^{" | jq -r '.acknowledged')
+    if [ "${acknowledged}" = "true" ]; then
+      log_info "- Successfully created alias .kibana"
+    else
+      log_fatal "Failed to create alias .kibana" "${resp}"
+    fi
+  else
+    log_fatal "Failed to check if alias .kibana exists" "${resp}"
+  fi
+
+}
+
+delete_opensearch_dashboards_index() {
+
+  os_dashboards_index=$(curl -sS -kL -u "${user}:${password}" -X GET "${os_url}"/_alias/.opensearch_dashboards | jq -r 'to_entries | .[0].key')
```

Comment:
Will there always only be one index pointed to by the alias?

Reply:
Yeah my understanding is that it always points to one index, and when it migrates to a new one it always re-points it to the new one. I checked this in some environments that have multiple indices and never saw an example where the alias pointed to anything other than one (the latest).

For example during this migration it will notice the `.kibana_1` index after the upgrade, migrate it to `.kibana_2` then re-point the alias to `.kibana_2`.

Reply:
It does not remove old indices after migrating. Not sure if it has some other mechanism for that purpose in the long term,  or if they are never cleaned up.

Reply:
Q. Should they be deleted or should we point the .kibana alias to them?

Reply:
I would say that we should delete them, as those indices will have system index permission issues if kept after the upgrade. 

Since the alias only points to the latest index, and after a migration it stops pointing to the old one, I would assume that it is the only one that it reads/writes from/to. So I'd say deleting them after cloning the latest to `.kibana_1` should be fine.

Reply:
Makes sense!